### ISSUE TITLE
Foreign keys disappear after merge for tables created with FOREIGN_KEY_CHECKS=0

### ISSUE THREAD

**Issue by arvidfm** [dolthub/dolt#7956]
If you create a table while `FOREIGN_KEY_CHECKS` is disabled, some of the foreign keys seem to disappear whenever a merge is next triggered (e.g. due to two simultaneous transactions).

Consider for instance:

```sql
SET FOREIGN_KEY_CHECKS=0;

-- set up other tables here

CREATE TABLE item_tasks (
  item_id INT,
  task_id varchar(100),
  user_id INT,
  FOREIGN KEY (item_id) REFERENCES items(id),
  FOREIGN KEY (task_id) REFERENCES tasks(id),
  FOREIGN KEY (user_id) REFERENCES user(id)
);

SET FOREIGN_KEY_CHECKS=1;
```

Then if you run two simultaneous transactions, e.g.:

```sql
[TX1] BEGIN;
[TX1] UPDATE tasks SET tasks.status = 'timed out';
[TX2] BEGIN;
[TX2] INSERT INTO user (username) VALUES ('test');
[TX2] COMMIT;
[TX1] COMMIT;
```

the foreign keys disappear from the `item_tasks` table (even though it wasn't modified in the transactions):

```
mysql> show create table item_tasks;
+------------+-----------------------------------+
| Table      | Create Table                      |
+------------+-----------------------------------+
| item_tasks | CREATE TABLE `item_tasks` (
  `item_id` int,
  `task_id` varchar(100),
  `user_id` int,
  KEY `item_id` (`item_id`),
  KEY `task_id` (`task_id`),
  KEY `user_id` (`user_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_bin |
+------------+-----------------------------------+
```

This is an issue for us as we have a very large set of migrations that are written with the assumption that `FOREIGN_KEY_CHECKS` is disabled to make it easier to modify the tables (with extra post-migration checks to verify that the constraints are still valid).

Here is a complete Go program that reproduces the issue every time for me:
```go
package main

import (
	"database/sql"

	"github.com/go-sql-driver/mysql"
)

const (
	dbAddr = "127.0.0.1:3306"
	dbUser = "root"
	dbPass = "toor"
	dbName = "dolt"
)

const setupStmt = `
DROP TABLE IF EXISTS item_tasks, items, tasks, user;

SET FOREIGN_KEY_CHECKS=0;

CREATE TABLE user (
  id INT PRIMARY KEY AUTO_INCREMENT,
  username VARCHAR(16)
);

CREATE TABLE tasks (
  id varchar(100) PRIMARY KEY,
  status ENUM('in progress', 'timed out', 'errored', 'completed')
);

CREATE TABLE items (
  id INT PRIMARY KEY AUTO_INCREMENT,
  user_id INT,
  FOREIGN KEY (user_id) REFERENCES user(id)
);

CREATE TABLE item_tasks (
  item_id INT,
  task_id varchar(100),
  user_id INT,
  FOREIGN KEY (item_id) REFERENCES items(id),
  FOREIGN KEY (task_id) REFERENCES tasks(id),
  FOREIGN KEY (user_id) REFERENCES user(id)
);

SET FOREIGN_KEY_CHECKS=1;
`

func main() {
	cfg := mysql.NewConfig()
	cfg.Net = "tcp"
	cfg.Addr = dbAddr
	cfg.User = dbUser
	cfg.Passwd = dbPass
	cfg.DBName = dbName
	cfg.MultiStatements = true

	conn := must(sql.Open("mysql", cfg.FormatDSN()))
	must(conn.Exec(setupStmt))

	tx1 := must(conn.Begin())
	must(tx1.Exec(`UPDATE tasks SET tasks.status = 'timed out'`))

	tx2 := must(conn.Begin())
	must(tx2.Exec(`INSERT INTO user (username) VALUES ('test')`))

	must1(tx2.Commit())
	must1(tx1.Commit())
}

func must[T any](t T, err error) T {
	if err != nil {
		panic(err)
	}
	return t
}

func must1(err error) {
	if err != nil {
		panic(err)
	}
}
```

**Comment by fulghum** [comment]
Thanks for the excellent repro case @arvidfm. I am able to repro the issue using your script, but I haven't been able to get a repro in our test framework yet. I have noticed that the FKs only get dropped when `tx2` is committed before `tx1`. I'll keep digging in and let you know what we find.

**Comment by fulghum** [comment]
Quick update on this one... I was able to find where this issue is coming from. Dolt foreign keys have a property of being "resolved" or not – an unresolved FK knows the names of the columns it references, but a resolved FK additionally knows the unique IDs of those columns, too. When FKs are added while `@@foreign_key_checks` is turned off, they are created as unresolved FKs, and they later get fully resolved to the unique column IDs the next time they are loaded when `@@foreign_key_checks` is enabled. In your repro code, one session/tx has the FKs loaded in a resolved state, and the other session/tx has the FKs loaded in an unresolved state. The code that merges two DB states has to merge the FK collections together, and it's a little tricky to merge the FKs when one set is resolved and the other is unresolved. 

I have a patch in my workspace that fixes the issue in this repro, but I still need to 1) dig around a little more and see if there's a better/cleaner way to handle this, 2) clean up this change and make sure it doesn't break any other test cases, and 3) write a repro test in our test framework (I'm still debugging off of your go script example, but I think I see all the moving parts for a repro in our test framework now). 

Thanks for helping us find this one! I'll stay focused on this one and will keep you updated with progress.

**Comment by arvidfm** [comment]
It's worth noting that it may or may not be some kind of race condition - I tried to record a trace of Dolt using `rr` while running our complete set of migrations to investigate why the FKs were disappearing, but it just wouldn't replicate while running in `rr`, maybe because the thread scheduling is different. (Though it does seem to replicate using my MWE above.) So maybe that could be part of the reason you're having issues replicating it in other environments.

P.S. Glad to hear you're getting some use out of the MWE, it was not very fun digging through our ~500 migration queries to figure out exactly what combination of queries (and in what order) was causing this to happen!

**Comment by fulghum** [comment]
This fix was reviewed and merged in. We'll run a release later today and close this issue once the release is out.

**Comment by arvidfm** [comment]
Amazing, thank you so much! I'll be sure to give this and the fixes for the other bugs you've looked at recently a spin with our migration set early next week!

**Comment by fulghum** [comment]
@arvidfm – The fix for this issue has been released in [Dolt 1.39.4](https://github.com/dolthub/dolt/releases/tag/v1.39.4). Major thanks for helping us find a repro for this tricky one! 🙏
